### PR TITLE
PIMOB-XXXX: Fix phone number validator behaviour

### PR DESCRIPTION
**Pull request by alex-ioja-yang-cko**
## Proposed changes

Simplified the data flow by removing unnecessary complications:
- removed concept of phone number specific textfield input
- improved Validator protocol to better provide expected functionality
- removed duplicate validators
- using Validator as validation source instead of mixing VC, VM & Validator...

Have added required business logic for Phone Number Validator to function as expected and updated its consumption to serve overall journey appropriately.

### CODE REVIEW

File: Checkout/Checkout/Source/Model/Country.swift [hunk 1]
```diff
@@ -284,6 +284,15 @@ public struct Country: Equatable {
       return nil
     }
   }
+    
+    public init?(dialingCode: String) {
```
Comment (deepesh-vasthimal-cko) [inline Checkout/Checkout/Source/Model/Country.swift:288]: So if merchant when initialising  Country  object with `+` you are removing it ?

File: Source/UI/Validator/GenericInputValidator.swift [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+class GenericInputValidator: Validator {
+
+    func shouldAccept(text: String) -> Bool {
```
[the remark this answers is not in the document]
Comment (deepesh-vasthimal-cko) [reply Source/UI/Validator/GenericInputValidator.swift:5]: Shouldnt this be a protocol with default implementation and then anyone else can use them else where ?

File: Source/UI/Validator/PhoneNumberValidator.swift [hunk 3]
```diff
@@ -1,11 +1,39 @@
 import Foundation
+import PhoneNumberKit
+import Checkout
 
 class PhoneNumberValidator: Validator {
     private let validator = PhoneValidator()
+    private let phoneKit = PhoneNumberKit()
+    private let validCharacterSet: CharacterSet = {
+        var validInputs = "+ "
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/Validator/PhoneNumberValidator.swift:9]: is the prefix always required ?

File: Source/UI/Validator/PhoneNumberValidator.swift [hunk 4]
```diff
@@ -1,11 +1,39 @@
...
-        guard let phoneNumber = (value as? Phone) else { return false }
-        return validator.validate(phoneNumber) == .success
+    func shouldAccept(text: String) -> Bool {
+        CharacterSet(charactersIn: text).isSubset(of: validCharacterSet) &&
+            text.count < Checkout.Constants.Phone.phoneMaxLength
+    }
+
+    func isValid(text: String) -> Bool {
+        do {
+            let formattedNumber = try phoneKit.parse(text, ignoreType: true)
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/Validator/PhoneNumberValidator.swift:21]: question what is the phoneKit.parse do ? here. Space formatting of phone numbers.

**Review by deepesh-vasthimal-cko — DISMISSED** [review]
Overall looks fine few questions. I can see some clean up which is good on PR. Would require test to make sure it works as expected.

File: Source/UI/Validator/GenericInputValidator.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Source/UI/Validator/GenericInputValidator.swift:5]: Yes, the `GenericInputValidator` doesn't have any validation logic, same as the 7 other validators it replaces :)

File: Source/UI/Validator/PhoneNumberValidator.swift (hunk 3 above)
Comment (alex-ioja-yang-cko) [reply Source/UI/Validator/PhoneNumberValidator.swift:9]: It is not required input, but it is valid input. The formatter will introduce it if not provided by user.

File: Source/UI/Validator/PhoneNumberValidator.swift (hunk 4 above)
Comment (alex-ioja-yang-cko) [reply Source/UI/Validator/PhoneNumberValidator.swift:21]: `phoneKit.parse` uses PhoneNumberKit to parse a String into a phone number. If it contains country code it uses the expected format for that country, if not it defaults to device locale (as we did previously)

_reference [PhoneNumberKit repo](https://github.com/marmelroy/PhoneNumberKit#usage)_

File: Tests/UI/BillingForm/View/BillingFormTextFieldViewTests.swift [hunk 5]
```diff
@@ -40,4 +40,41 @@ class BillingFormTextFieldViewTests: XCTestCase {
...
+    func testUpdateStyleFormatsPhoneNumber() {
+        var style = DefaultBillingFormPhoneNumberCellStyle()
+        style.textfield.text = "01206123123"
+        let view = BillingFormTextFieldView()
+        view.update(style: style, type: .phoneNumber(style), tag: 0)
+        
+        XCTAssertEqual(view.textField.text, "+44 1206 123123")
+    }
+    
+    func testShouldEndEditingIsFormattingDisplay() {
```
Comment (deepesh-vasthimal-cko) [inline Tests/UI/BillingForm/View/BillingFormTextFieldViewTests.swift:53]: 👍

File: Tests/UI/BillingForm/View/BillingFormTextFieldViewTests.swift [hunk 6]
```diff
@@ -40,4 +40,41 @@ class BillingFormTextFieldViewTests: XCTestCase {
...
+        let testTextField = UITextField()
+        testTextField.text = "01206123123"
+        let view = BillingFormTextFieldView()
```
Comment (deepesh-vasthimal-cko) [inline Tests/UI/BillingForm/View/BillingFormTextFieldViewTests.swift:56]: is it possible to declare testTextField, view ( BillingFormTextFieldView() ) and style (DefaultBillingFormPhoneNumberCellStyle() ) externally to be reused in different test cases ?

**Review by deepesh-vasthimal-cko — APPROVED** [review]
Just one minor comment. Otherwise looks good.

File: Tests/UI/BillingForm/View/BillingFormTextFieldViewTests.swift (hunk 6 above)
Comment (alex-ioja-yang-cko) [reply Tests/UI/BillingForm/View/BillingFormTextFieldViewTests.swift:56]: Each test case should build its own environment (as long as not overly complex - not the case here), I fundamentally disagree with the approach you're suggesting.

File: Source/UI/Theme/Generic/Theme+TextField.swift [hunk 7]
```diff
@@ -15,7 +15,7 @@ public extension Theme {
         public var text: String
         public var isSupportingNumericKeyboard = true
         public var textAlignment: NSTextAlignment = .natural
-        public var height: Double = 30
+        public var height: Double = 56
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/Theme/Generic/Theme+TextField.swift:18]: 👍

**Review by deepesh-vasthimal-cko — APPROVED** [review]

**Pull request merged**